soundcard/pulseaudio.py

```diff
@@ -333,45 +332,78 @@ class _Recorder(_Stream):
         self._pulse._pa_stream_connect_record(self.stream, self._id.encode(), bufattr, _pa.PA_STREAM_ADJUST_LATENCY)
 
-    def record(self, numframes):
-        """Record some audio data.
-
-        The data will be returned as a `frames × channels` float32
-        numpy array.
-
-        This function will wait until `numframes` frames have been
-        recorded. However, the audio backend holds the final authority
-        over how much audio data can be read at a time, so the
-        returned amount of data will often be slightly larger than
-        what was requested. The amount of buffering can be controlled
-        through the blocksize of the recorder object.
-
-        """
-
-        captured_frames = 0
-        captured_data = []
+    def _record_chunk(self):
         data_ptr = _ffi.new('void**')
         nbytes_ptr = _ffi.new('size_t*')
-        while captured_frames < numframes:
+        while True:
             readable_bytes = self._pulse._pa_stream_readable_size(self.stream)
             if readable_bytes > 0:
                 data_ptr[0] = _ffi.NULL
                 nbytes_ptr[0] = 0
                 self._pulse._pa_stream_peek(self.stream, data_ptr, nbytes_ptr)
                 if data_ptr[0] != _ffi.NULL:
-                    chunk = numpy.fromstring(_ffi.buffer(data_ptr[0], nbytes_ptr[0]), dtype='float32')
+                    chunk = np.fromstring(_ffi.buffer(data_ptr[0], nbytes_ptr[0]), dtype='float32')
                 if data_ptr[0] == _ffi.NULL and nbytes_ptr[0] != 0:
-                    chunk = numpy.zeros(nbytes_ptr[0]//4, dtype='float32')
+                    chunk = np.zeros(nbytes_ptr[0]//4, dtype='float32')
                 if nbytes_ptr[0] > 0:
                     self._pulse._pa_stream_drop(self.stream)
-                    captured_data.append(chunk)
-                    captured_frames += len(chunk)/self.channels
+                    return chunk
             else:
                 time.sleep(0.001)
-        return numpy.reshape(numpy.concatenate(captured_data), [-1, self.channels])
+
+    def record(self, numframes=None):
+        """Record a block of audio data.
+
+        The data will be returned as a frames × channels float32 numpy array.
+        This function will wait until numframes frames have been recorded.
+        If numframes is given, it will return exactly `numframes` frames,
+        and buffer the rest for later.
+
+        If numframes is None, it will return whatever the audio backend
+        has available right now.
+        Use this if latency must be kept to a minimum, but be aware that
+        block sizes can change at the whims of the audio backend.
+
+        If using `record` with `numframes=None` after using `record` with a
+        required `numframes`, the last buffered frame will be returned along
+        with the new recorded block.
+        (If you want to empty the last buffered frame instead, use `flush`)
+        """
+        if not numframes:
```

Comment:
this should be `if numframes is None`, so as to not confuse `numframes=0` with `numframes=None`.